### PR TITLE
Pipeline: Reduce CPU usage of `WaitReactor`

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?

Issue Number: ref #6518

Problem Summary:

#### optimize
- add function `Operator::isWaitable` to reduce unnecessary `Operator::await` calls.
- refine `ExchangeReceiver::receive` to reduce the overheads of `std::function`.
- rename `isReadyForWite` to `isWritable`
- move `Spinner` to `WaitReactor` and inline some functions.
- use `_mm_pause` for `x86_64` in `WaitReactor`
- reduce high frequency `MemoryTrackerPtr::~shared_ptr()` of `TaskHelper::TRACE_MEMORY(task)`
YbPEAcGvNz
- reduce high frequency `exec_status.isCancelled()` in `Operator::await`
3o0aJao2wn
- reduce high frequency `MemoryTrackerPtr::get()` of `TaskHelper::TRACE_MEMORY(task)`
LAuCDSXH6d
- remove useless `runtime_assert` in `Task::assertNormalStatus`
- reduce the overheads of lambda in `EventTask::doTaskAction`.
38H0DHxiKp

#### result
the cpu usage of `WaitReactor` from 12% to 4.4% (tpch q1)

### Check List

Tests 

- [x] compute side tsan passed
  - https://ci.pingcap.net/blue/organizations/jenkins/tiflash-sanitizer-daily/detail/tiflash-sanitizer-daily/1095/pipeline
- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by SeaRise** [comment]
> /run-sanitizer-test tsan

https://ci.pingcap.net/blue/organizations/jenkins/tiflash-sanitizer-daily/detail/tiflash-sanitizer-daily/1095/pipeline

File: dbms/src/Flash/Pipeline/Exec/PipelineExec.cpp [hunk 1]
```diff
@@ -46,6 +46,41 @@ namespace DB
...
+
+PipelineExec::PipelineExec(
+    SourceOpPtr && source_op_,
+    TransformOps && transform_ops_,
+    SinkOpPtr && sink_op_)
+    : source_op(std::move(source_op_))
+    , transform_ops(std::move(transform_ops_))
+    , sink_op(std::move(sink_op_))
+{
+#define FILL_AWAITABLE(op)                                                   \
```
Comment (yibin87) [inline dbms/src/Flash/Pipeline/Exec/PipelineExec.cpp:72]: use inline method to replace macro? And can add comments to limit its usage.

File: dbms/src/Flash/Mpp/MPPTunnel.cpp [hunk 2]
```diff
@@ -134,23 +135,22 @@ void MPPTunnel::close(const String & reason, bool wait_sender_finish)
...
+            need_close_tunnel = true;
             break;
-        }
         case TunnelStatus::Finished:
             return;
         default:
             RUNTIME_ASSERT(false, log, "Unsupported tunnel status: {}", static_cast<Int32>(status));
         }
     }
+    if (need_close_tunnel)
```
[the remark this answers is not in the document]
Comment (SeaRise) [reply dbms/src/Flash/Mpp/MPPTunnel.cpp]: make sense, reverted

File: dbms/src/Flash/Pipeline/Exec/PipelineExec.cpp (hunk 1 above)
Comment (SeaRise) [reply dbms/src/Flash/Pipeline/Exec/PipelineExec.cpp]: ok, updated.

File: dbms/src/Operators/UnorderedSourceOp.h [hunk 3]
```diff
@@ -26,6 +26,7 @@ namespace DB
 /// Read blocks asyncly from Storage Layer by using read thread,
 /// The result can not guarantee the keep_order property
 class UnorderedSourceOp : public SourceOp
+    , public Awaitable
```
Comment (windtalker) [inline dbms/src/Operators/UnorderedSourceOp.h:29]: Is inheriting from multiple classes aims at avoiding virtual function call?(I mean why not add a virtual function `isAwaitable()`)

Comment (SeaRise) [reply dbms/src/Operators/UnorderedSourceOp.h:29]: Yes, this implementation is just an option. 
Adding `isAwaitable()` means that the awaitable operator needs to override two methods.

Comment (SeaRise) [reply dbms/src/Operators/UnorderedSourceOp.h]: updated to ·`isAwaitable()`

**Review by yibin87 — APPROVED** [review]
LGTM

**Review by windtalker — APPROVED** [review]
LGTM

**Pull request merged**